src/gsy_e/models/market/forward.py

```diff
@@ -0,0 +1,138 @@
+            return
+
+        super().__init__(bc=bc, notification_listener=notification_listener,
+                         readonly=readonly, grid_fee_type=grid_fee_type,
+                         grid_fees=grid_fees, name=name)
+
+    @staticmethod
+    @abstractmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block starts."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block ends."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_market_slot_duration(current_time: DateTime, _config) -> duration:
+        """Return duration of market slots inside the market block."""
+
+    def create_future_market_slots(self, current_market_time_slot: DateTime,
+                                   config: "SimulationConfig") -> None:
+        if not ConstSettings.ForwardMarketSettings.ENABLE_FORWARD_MARKETS:
+            return
+        self._create_future_market_slots(self._get_start_time(current_market_time_slot),
+                                         self._get_end_time(current_market_time_slot), config)
+
+
+class DayForwardMarket(ForwardMarketBase):
+    """Day forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(days=1)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(weeks=1, days=1, hours=-1)
+
+    @staticmethod
+    def _get_market_slot_duration(_current_time: DateTime, _config) -> duration:
+        return duration(hours=1)
+
+
+class WeekForwardMarket(ForwardMarketBase):
+    """Week forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        days_until_next_monday = 7 - (current_time.day_of_week - 1)
+        return current_time.set(hour=0, minute=0).add(days=days_until_next_monday)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(years=1)
+
+    @staticmethod
+    def _get_market_slot_duration(_current_time: DateTime, _config) -> duration:
+        return duration(weeks=1)
+
+
+class MonthForwardMarket(ForwardMarketBase):
+    """Month forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(day=1, hour=0, minute=0).add(months=1)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(day=1, hour=0, minute=0).add(years=2)
+
+    @staticmethod
+    def _get_market_slot_duration(current_time: DateTime, _config) -> duration:
+        number_of_days_month = monthrange(current_time.year, current_time.month)[1]
+        return duration(days=number_of_days_month)
```

Comment:
pendulum.DateTime is smart enough to know about number of days in each month, thus:
```python
return duration(months=1)
```

Reply:
I do not really believe that, because different months have different number of days. What is probably misleading you here is the `current_time`, which is not really the wall clock time, but the `current_simulation_time`, which will coincide with the wall clock for CNs, but will not for simulations (e.g. I can start a simulation now for February 2019, and the function will return 31 days, even though the intended result of the method would be 28/29 days)

Reply:
![Uploading Screenshot from 2022-07-19 09-54-09.png…]()

Reply:
@mrsaemir I am not sure about that, check out this output from ipython:

```
In [1]: from pendulum import duration

In [2]: d = duration(months=1)

In [3]: d.days
Out[3]: 30
```

Reply:
I think that using `__add__` (`+`) between `DateTime` and `duration` takes days into account:
```py
> january
DateTime(2021, 1, 1, 0, 0, 0)
> january + duration(months=1)
DateTime(2021, 2, 1, 0, 0, 0)
> february = DateTime(2021, 2, 1)
> february.days_in_month
28
> february + duration(months=1)
DateTime(2021, 3, 1, 0, 0, 0)
>
DateTime(2021, 3, 1, 0, 0, 0)
```

Reply:
Also, this is part of the implementation for .add() method for pendulum.DateTime objects:
```python
    if months:
        month += months
        if month > 12:
            year += 1
            month -= 12
        elif month < 1:
            year -= 1
            month += 12

    day = min(DAYS_PER_MONTHS[int(is_leap(year))][month], dt.day)
    ```

Reply:
What you suggest implies though that this duration will only be used in adding / subtracting to datetime values, which I am not sure is the case here. Should be the case indeed, however it needs to be checked whether it indeed works the same way as with days (which we know that works even if the duration is converted to days int - which I suspect is the reason that it is implemented like this here). 

If it indeed is the case then all good, however I wonder whether you noticed an issue somewhere due to this calculation. Regardless of that I do not mind if this discussion is theoretical, or more focused in order to use pendulum here too, however if this really causes an issue it would be great to also point it out to see what causes it.

Reply:
> Edit: meanwhile Pierpaolo confirmed my suspicion:
> 
> ```
> future_time_slot.add(minutes=self._get_market_slot_duration(future_time_slot, config).total_minutes())
> ```
> 
> this will not work with `duration(month=1)` because this will always return 30 days and will not take into account the future_time_slot.

Regarding this, my suggested solution would be to avoid using `.add()` and use
```py
future_time_slot + self._get_market_slot_duration(future_time_slot, config)
```
This will take the actual date of `future_time_slot` and add the correct duration to it (e.g. 1 month, automatically adjusting the number of days).

Reply:
So, to recap and resolve this conversation, are we going to return `duration` objects in `_get_market_slot_duration` methods and use the `+` operator instead of the `.add` method?
@spyrostz @fievelk

Reply:
I've just pushed one commit that changes that logic: 54ec1320d. If this is not optimal, please feel free to revert it in the next days.